### ISSUE TITLE
CRAN release?

### ISSUE THREAD

**Issue by cdeterman** [catboost/catboost#439]
Problem: {No CRAN released version}

Is there a particular reason that there is no version of catboost released to the CRAN ecosystem for R?

**Comment by annaveronika** [comment]
It would be great to make a CRAN release, but we need help with that.

**Comment by cdeterman** [comment]
@annaveronika what help do you need?  I maintain multiple R packages and have helped others publish their packages.  I would be happy to be of assistance.

**Comment by annaveronika** [comment]
https://tech.yandex.com/catboost/doc/dg/concepts/installation-docpage/ - here is the instruction, how to install the package.

We either provide the binaries or it is also possible to build from source, but we had many problems of people trying to build from source on windows, so the easiest way to install is to get the binaries.

We normally build using our own build system. But we also provide makefiles for python and cmdline. 
https://github.com/catboost/catboost/tree/master/make

I don't think we can upload the binary of the R package to CRAN.

Plus you need to have in mind that we are going to add GPU support to the package.

Given all this, what are the requirements and the first steps for catboost?

**Comment by cdeterman** [comment]
Ideally we want to get the windows builds to work.  You are correct that binaries aren't allowed for submission to CRAN.  Strictly speaking Windows support isn't required but obviously something we should try to sort out.  Are there any dependencies that aren't normally installed on Windows or is the package self contained?  I tried installing it on my own Windows machine and ran in to a list of problems (truncated for space).


>Info: Attention! Using system user-defined compiler: None (check CC and CXX env vars).
Warn[-WNoInput]: in $S/contrib/libs/openssl/ya.make: macro for $S/contrib/libs/openssl/asm/windows/aes-x86_64.masm didn't use it as input
...
Error[-WNoOutput]: in $S/contrib/libs/openssl/ya.make: macro 9010:SRCSmasm=(x86_64cpuid.masm) resulted in no outputs, can't add to graph
...
Configure error (use -k to proceed)
make: *** [all] Error 1
Warning: running command 'make --no-print-directory -f "Makefile.win"' had status 2
ERROR: compilation failed for package 'catboost'



Is there something unique about your build system?

Does the package 'always' build successfully on Linux/OSX systems?  I don't see anything with respect to a continuous integration system that could be of use for this (e.g. Travis CI).

Regarding GPU support, that shouldn't be problem.  I am very familiar with this as I am the author of the `gpuR` package.  There are a few moving parts but it ultimately shouldn't be a major problem.

**Comment by annaveronika** [comment]
@cdeterman Sorry for a delay in response. We'll come back to you with more details later.

**Comment by timchurches** [comment]
Not sure how feasible this would be, but it would be great if CatBoost could also be made to work with the ViennaCL library that @cdeterman uses in his gpuR package, mainly because that would mean it could make use of the AMD GPUs built into many Macbook Pros (and iMacs), which support double-precision maths (unlike the Intel Iris GPUs that lower-end Mackbooks use).

**Comment by jameslamb** [comment]
@annaveronika are you open to creating a GitHub milestone enumerating what needs to be done to get to CRAN?

I found working on `LightGBM` and other projects that just listing out the work (however much there is) makes it more friendly for outside contributors to help feel that they're moving the project forward.

I would be willing to help contribute getting past some of the more esoteric CRAN stuff.

**Comment by annaveronika** [comment]
https://github.com/catboost/catboost/pull/715 - here is a pr that is preparation for CRAN.
@ws171913 is working on adding CatBoost to CRAN, please look on the PR and comment if you have any questions.

**Comment by jameslamb** [comment]
Awesome thanks!

**Comment by Avichi** [comment]
Folks I was able to install CatBoost using the GitHub cloning on local PC on Windows 10 64x, and then pointing the command to install it from the folder (Make sure the folder has access to the folders "permission wise"). The GPU is working on Windows 10 and am testing the catBoost parameters on GPU. It works great  on Windows 10 R 3.6.1 on R-Studio IDE Thank you @annaveronika

**Comment by Steviey** [comment]
+1

**Comment by lorenzwalthert** [comment]
May I ask why the *in progres* label was removed? I also see #715. A lot of people only use packages that are distributed via CRAN. Thanks for this package.

**Comment by t-wojciech** [comment]
Indeed, publishing the package on CRAN would be great news. I look forward to it!

**Comment by maksymiuks** [comment]
Would be great to see this package on CRAN. Especially because other packages that may want to import or cooperate with Yours cannot do it since it's impossible to put catboost in a valid DESCIRPTION file of other CRAN package.

**Comment by david-cortes** [comment]
@jameslamb : another potential obstacle for getting accepted at CRAN, and which would be good to solve anyway, is that the documentations of the package's functions do not describe the return values.

**Comment by jameslamb** [comment]
I know how to do that! It's something I could help with :grinning:

**Comment by kizill** [comment]
> I know how to do that! It's something I could help with 😀

Sounds really great! If you'll need some help - feel free to contact me here or via Telegram @kizill.

**Comment by david-cortes** [comment]
@kizill : Something very helpful for anyone working on potential CRAN obstacles would be to provide some build system that would allow compiling it from source within R. Right now if I try to install the R package from a locally-downloaded repository it will download the precompiled library instead of building it.

Would like to point out the obvious too:
* R packages expect the files that need compilation to be under a folder called `/src` in the R package's folder. Would be nice to provide a script that would copy the files there from the repository's root. From a look at the directory tree I couldn't figure out which files need to be compiled to have a working library, which are optional, and which are not R-related.
* CRAN uses the GNU make system. It expects a `Makevars` file, which is similar to but not the same as the standard makefiles used by catboost. `Makevars` can also be auto-generated through a `configure` script. The `configure` script should also be in the R package directory root.
* Messages printed to `stdout` and `stderr` need to use R's own IO functions. This can be as easy as `#define`'ing `printf` -> `Rprintf` (or `cout` -> `Rcout` if using Rcpp and C++ streams).
* Some of the examples and tests are currently failing.

**Comment by david-cortes** [comment]
@kizill : I also notice that the R wrapper doesn't use R' own exporting and registration methods for the C/C++ functions (`R_registerRoutines` + `R_useDynamicSymbols` + `.register=TRUE`), and the windows version uses instead `__declspec(dllexport)`, which would most definitely fail CRAN automated checks.

Is  there some particular reason why the library is avoiding R's own export/registration routines? Would something break if they are added?

**Comment by jameslamb** [comment]
A new project, `{bonsai}`, was recently released with support from RStudio.

Since it might be help maintainers to understand the impact of the R package note being on CRAN.

https://www.tidyverse.org/blog/2022/06/bonsai-0-1-0/#other-notes

> Unfortunately, the catboost R package is not on CRAN, so we’re not able to add support for the package for now.

**Comment by joeycouse** [comment]
Would love to see Catboost on CRAN! If Catboost was on CRAN it would be much more feasible for other developers to use it in other packages and provide integration with the `tidymodels` framework.

**Comment by Sinansi** [comment]
Any update on this issue? Is CatBoost going to be on CRAN?

**Comment by andrey-khropov** [comment]
> Any update on this issue? Is CatBoost going to be on CRAN?

https://github.com/catboost/catboost/milestone/2. Switching CatBoost to CMake-based build in release 1.2 paved the way to CRAN release, but there is still more work to do. If you can contribute you might speed up the process.

**Comment by larry77** [comment]
Any news about the catboost release on CRAN? I see interest from different machine learning frameworks on R (tidymodels and mlr3).

**Comment by andrey-khropov** [comment]
> Any news about the catboost release on CRAN? I see interest from different machine learning frameworks on R (tidymodels and mlr3).

I expect CatBoost 1.3 that should be on CRAN in two to three months.

**Comment by jrosell** [comment]
Any news on that?

**Comment by andrey-khropov** [comment]
> Any news on that?

work in progress

**Comment by david-cortes** [comment]
Couple requirements from CRAN to point out just in case:
* Examples and tests should not create nor delete any files in the current working directory or in some user directory. For files that are created during examples, one should use the designated temporary directory, which can be retrieved through `tempdir()`.
* Examples and tests should not use more than 2 parallel threads.
* All bundled dependencies should be credited if their license requires it. So for example, copyright notices from MIT/BSD licenses should be placed under a file `inst/COPYRIGHTS`. You might also want to add copyright holders of those dependencies in the `DESCRIPTION` file.

**Comment by davidkane9** [comment]
Adding another voice to those who would like to see catboost on CRAN. I would love to use catboost for my [introductory data science course](https://bootcamp.davidkane.info/), but I am hesitant about having beginners (try to) install non-CRAN packages.

**Comment by jrosell** [comment]
> Adding another voice to those who would like to see catboost on CRAN. I would love to use catboost for my [introductory data science course](https://bootcamp.davidkane.info/), but I am hesitant about having beginners (try to) install non-CRAN packages.

For this purpose, you may want to use a Kaggle notebook at the moment. There are some runable R examples using Catboost out there.

**Comment by ekianjo** [comment]
> > Any news on that?
> 
> work in progress

Do you have any new ETA at the moment?

**Comment by andrey-khropov** [comment]
> > > Any news on that?
> > 
> > 
> > work in progress
> 
> Do you have any new ETA at the moment?

We’ve had multiple issues delaying release 1.3. I expect it to be released at the end of April. Sorry for the wait.

**Comment by larry77** [comment]
Any news about this?

**Comment by andrey-khropov** [comment]
> Any news about this?

We are still working on it.

**Comment by larry77** [comment]
Any news?

**Comment by SamChakwera** [comment]
Still nothing?

**Comment by larry77** [comment]
Hellooooo!

**Comment by Sinansi** [comment]
Any updates on this issue?

**Comment by frankiethull** [comment]
dear R programmers: 

You can use catboost in R without CRAN (see catboost's install instructions). 

You can also (as of June 2025) use catboost with `bonsai`, the `tidymodels` extension package for tree-based models (see new [engines](https://bonsai.tidymodels.org/news/index.html#new-engines-0-4-0)). 

I am assuming that many are here for `tidymodels` support. In the past, the team did not support non-CRAN packages. But due to the many many many requests for catboost, Emil bent the rules (just for you all!). 

TLDR: I think R programmers can use catboost & not bother team about a CRAN release, unless, of course, you have a solution that can help them submit to CRAN 🙏🏻

**Comment by larry77** [comment]
Well, at least for me it is now problem solved ☺️
